src/System.Text.Json/System/Text/Json/JsonWriter.cs

```diff
@@ -6,193 +6,592 @@
+            WriteItemSeperator();
+            _firstItem = false;
+            WriteSpacing();
+            WriteGuid(value);
         }
 
-        private void ConsideComma()
+        /// <summary>
+        /// Write a null value into the current array.
+        /// </summary>
+        public void WriteNull()
         {
-            if (_wroteListItem) {
-                _formatter.Append(',');
-                WriteNewline();
-            }
+            WriteItemSeperator();
+            _firstItem = false;
+            WriteSpacing();
+            WriteAscii(JsonConstants.NullValue);
+        }
+
+        /// <summary>
+        /// Write an undefined value into the current array.
+        /// </summary>
+        public void WriteUndefined()
+        {
+            WriteItemSeperator();
+            _firstItem = false;
+            WriteSpacing();
+            WriteAscii(JsonConstants.UndefinedValue);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void WriteStartAttribute(string name)
+        {
+            WriteItemSeperator();
+            _firstItem = false;
+
+            WriteSpacing();
+            WriteQuotedString(name);
+            WriteChar(JsonConstants.KeyValueSeperator);
+
+            if (_prettyPrint)
+                WriteChar(JsonConstants.Space);
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private void WriteNewline()
+        private void WriteChar(char value)
         {
-            if (PrettyPrint) {
-                _formatter.Append('\n');
+            var output = _output;
+            if (UseFastUtf8)
+            {
+                if (output.Buffer.Length < 1)
+                    output.Enlarge();
+
+                output.Buffer.DangerousGetPinnableReference() = (byte)value;
+                output.Advance(1);
             }
+            else if (UseFastUtf16)
+            {
+                if (output.Buffer.Length < 2)
+                    output.Enlarge(2);
+
+                unsafe
+                {
+                    Unsafe.AsRef<char>(Unsafe.AsPointer<byte>(ref output.Buffer.DangerousGetPinnableReference())) = value;
+                }
+
+                output.Advance(2);
+            }
+            else
+            {
+                unsafe
+                {
+                    // Slow path, if we are dealing with non-invariant.
+                    ReadOnlySpan<char> chSpan = new ReadOnlySpan<char>(&value, 1);
+                    Write(chSpan);
+                }
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void WriteQuotedString(string value)
+        {
+            WriteChar(JsonConstants.Quote);
+            // TODO: We need to handle escaping.
+            Write(value.AsSpan());
+            WriteChar(JsonConstants.Quote);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void WriteNumber(long value)
```

Comment:
All these write methods implement a loop that is very similar to what System.Text.Formatting component does. Why cannot these methods be reusable? Can we somehow make them reusable so that every developer who wants to create a writer does not have to write the same boilerplate.